tracing.go

```diff
@@ -166,21 +166,28 @@ func (c *Connection) startOutboundSpan(ctx context.Context, serviceName, methodN
 // when the outbound call is initiated. The tracing API is used to serialize the span
 // into the application `headers`, which will propagate tracing context to the server.
 // Returns modified headers containing serialized tracing context.
+//
+// Sometimes caller pass a shared instance of the `headers` map, so instead of modifying
+// it we clone it into the new map (assuming that Tracer actually injects some tracing keys).
 func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]string) map[string]string {
 	span := response.span
 	if span == nil {
 		return headers
 	}
-	if headers == nil {
-		headers = make(map[string]string)
-	}
-	carrier := tracingHeadersCarrier(headers)
+	newHeaders := make(map[string]string)
+	carrier := tracingHeadersCarrier(newHeaders)
 	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err != nil {
```

Comment:
It's hard to see what's actually going on here because the `carrier` is actually the one operating on the headers map. It seems more reasonable for the tracer to always make a new headers map on `Inject` rather than try to understand all the implicit mutations that are going on. It's probably outside the scope of this PR, but I think it should be the responsibility of `Tracer().Inject()` to create the new map if it's going to make these assumptions.